optionally add MDTags to reads with `transform`

ryan-williams opened the pull request:
fixes #738 

The entire reference is collected on to the driver and broadcast to workers.

I thought about doing some fancier region-join stuff, but then just decided to do a first pass this way.

I tested it on [`bqsr.sam` from this repo](https://github.com/bigdatagenomics/adam/blob/207eebaf3dbf6f01e5b8398b2af388ee70c862aa/adam-cli/src/test/resources/bqsr1.sam) as follows:

```
wget ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/vertebrates_mammals/Homo_sapiens/GRCh37/Primary_Assembly/assembled_chromosomes/FASTA/chr22.fa.gz
gunzip chr22.fa.gz

# rename the contig to "22" to match bqsr.sam
cat <(head -n 1 chr22.fa | pe 's/^>/>22 /') <(tail -n +2 chr22.fa) > chr22.fa.tmp
mv chr22.fa{.tmp,}

# Make a version of bqsr.sam with no MD tags
perl -pe 's/\tMD:Z:[^\s]+//' adam-cli/src/test/resources/bqsr1.sam > adam-cli/src/test/resources/bqsr1.stripped.sam

# Run transform! input: "stripped" .sam with no MDTags; output: "restored" .sam with MDTags.
bin/adam-submit --driver-memory 4g --conf spark.kryoserializer.buffer.max=256m -- transform adam-cli/src/test/resources/bqsr1.stripped.sam -add_md_tags chr22.fa -single adam-cli/src/test/resources/bqsr1.restored.sam

# This exits 0; the MD tags in the original and "restored" files are identical.
diff -q <(pcregrep -o 'MD:Z:[^\s]+' adam-cli/src/test/resources/bqsr1.sam) <(pcregrep -o 'MD:Z:[^\s]+' adam-cli/src/test/resources/bqsr1.restored.sam)
```

fnothaft commented:
Nice! Would you like to get this in to 0.17.1? I will make a review pass now.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala:
```diff
@@ -19,7 +19,7 @@ package org.bdgenomics.adam.rdd.read
 import java.io.StringWriter
 
-import htsjdk.samtools.{ SAMFileHeader, SAMTextHeaderCodec, SAMTextWriter, ValidationStringency }
+import htsjdk.samtools.{ TextCigarCodec, SAMFileHeader, SAMTextHeaderCodec, ValidationStringency, SAMTextWriter }
```
Small nit: at this point, might it not make sense to just do `import htsjdk.samtools._`? I think the general guideline we've been using is to import the whole package if we are using more than 4 items from the package.

ryan-williams replied:
haha funny you should ask; I [just complained to JetBrains](https://youtrack.jetbrains.com/issue/IDEA-144185) about IntelliJ reverting my disabling of the wildcard-imports when it exceeds 4, the preferences-search making it hard to find that setting, etc.

I'm fine to wildcard it if that's what you prefer.

fnothaft replied:
I actually agree with you RE: not wildcard importing packages. My personal preference would be to newline it at the commas though, and to alphabetize the classes that are imported (I prefer to keep the imports alphabetized, because it generally keeps diffs smaller).

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala:
```diff
@@ -285,6 +285,77 @@ class AlignmentRecordRDDFunctions(rdd: RDD[AlignmentRecord])
+    )
+
+  def adamAddMDTags(referenceFragments: RDD[NucleotideContigFragment]): RDD[AlignmentRecord] = {
+    val collectedRefMap =
```
As future cleanup, we might want to make this object (the collectedRefMap) extend [ReferenceFile](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/util/ReferenceFile.scala), and then change `getRefSeq` to `ReferenceFile.extract`. Then we could use this with both 2Bit files and etc.

ryan-williams commented:
Thanks @fnothaft, no particular rush; this impl may be too inefficient for prime-time.

I updated the description with some info about a one-contig test I did locally, fjyi.

fnothaft commented on adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala:
```diff
@@ -285,6 +285,77 @@ class AlignmentRecordRDDFunctions(rdd: RDD[AlignmentRecord])
+        referenceEnd = referenceStart + referenceLength
+        refSeq = getRefSeq(contigName, referenceStart, referenceEnd)
+        cigar = TextCigarCodec.decode(read.getCigar)
+        mdTag = MdTag(read.getSequence, refSeq, cigar, referenceStart)
```
Nice! This makes it _way_ simpler than I'd originally thought it'd be.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/856/<h2>Build result: FAILURE</span></h2>[...truncated 16 lines...]First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.10,1.2.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.2.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.3.1,centosADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
Hm:

```
[ERROR] /home/jenkins/workspace/ADAM-prb/HADOOP_VERSION/2.3.0/SCALAVER/2.10/SPARK_VERSION/1.2.1/label/centos/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala:310: error: could not find implicit value for parameter param: org.apache.spark.AccumulatorParam[Long]
[ERROR]     val mdTagsAdded = sc.accumulator(0L, "MDTags Added")
[ERROR]                                     ^
```

Seems to uniformly pass on Spark > 1.2 and fail on <= 1.2:

![](http://cl.ly/image/143q3j2N2d41/Screen%20Shot%202015-08-20%20at%209.28.49%20PM.png)

fnothaft commented:
What does the performance look like on the single contig? Personally, I would expect it to be pretty reasonable. The MDTag code is the same code that we rely on in the INDEL realigner, which is fairly zippy.

fnothaft commented:
The `AccumulatorParam[Long]` definitely exists in 1.2.0...: https://spark.apache.org/docs/1.2.0/api/scala/index.html#org.apache.spark.SparkContext$$LongAccumulatorParam$

ryan-williams commented:
I guess by "performance" I meant:
* the driver needs enough memory to collect the full reference
* we're broadcasting the full reference to all nodes
* I seemingly had to increase my `spark.kryoserializer.buffer.max` to run on chr22, which is ~50m as a FASTA (probably more as `NucleotideContigFragment`s.
  * It defaults to `64m`, I set it to `256m` as a guess and that worked.
  * I'm worried that I'd have to make that large enough for the whole reference...
  * Will we run into the "blocks must be <2GB" issue? [SPARK-6190](https://issues.apache.org/jira/browse/SPARK-6190)

ryan-williams commented:
Thoughts on how I should quantify/measure the performance? I can try to run it on a whole BAM on a cluster, and look at some of the Spark-reported numbers, but I'm not sure that will cover some of the things I'm really curious about.

fnothaft commented:
> I guess by "performance" I meant:
>
> * the driver needs enough memory to collect the full reference

Ah, gotcha. That's part of what I was thinking RE: refactoring around the `ReferenceFile` trait, since we could broadcast 2Bit files, which are fairly compact in memory (at the least, @laserson and I have both successfully used whole genome 2BIT files as broadcast objects; for hg19, the 2bit file is ~800MB), at the possible cost of increased region extraction costs.

FWIW, we have a few fairly large broadcast objects in the ADAM codebase (e.g., for BQSR) so adding another large broadcast doesn't trouble me too much.

> Thoughts on how I should quantify/measure the performance? I can try to run it on a whole BAM on a cluster, and look at some of the Spark-reported numbers, but I'm not sure that will cover some of the things I'm really curious about.

I was personally just looking for ballpark numbers. I do agree that the whole genome broadcast object is most likely to be the sticking point, so running that test seems reasonable.

Actually, I've got the high coverage NA12878 sitting on a cluster, along with the 1000G custom reference. I'll build and run this.

ryan-williams commented:
### Accumulators
I feel like it has to do with the `implicit` `LongAccumulatorParam` object, [which showed up in 1.3.0](https://github.com/apache/spark/blob/v1.3.0/core/src/main/scala/org/apache/spark/Accumulators.scala#L267-L270) but [was not in 1.2.1](https://github.com/apache/spark/blob/v1.2.1/core/src/main/scala/org/apache/spark/Accumulators.scala).

I'm not totally clear on why that is necessary, but I'll dig; let me know if you have thoughts about it.

### BroadcastRegionJoin

I just realized that I basically reimplemented `BroadcastRegionJoin` here; I'm going to look into reworking this to use that. lmk if that doesn't make sense.

I'm also going to look into an implementation that does a `ShuffleRegionJoin`, I have decided on an approach that should work for that.

### Testing on WGS
Is it easy for you to point me at the NA12878 and reference you speak of?

Should [eggo](https://github.com/bigdatagenomics/eggo) be useful to us here? cc @laserson

fnothaft commented:
> Testing on WGS
>
> Is it easy for you to point me at the NA12878 and reference you speak of?

For some _blasted_ reason, 1000G pulled down the high coverage WGS copy of NA12878. _Howaaaaaayver_, UCSC has a copy of it up at https://s3-us-west-2.amazonaws.com/bd2k-test-data/NA12878.mapped.ILLUMINA.bwa.CEU.high_coverage_pcr_free.20130906.bam. The reference that file is aligned to is ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/human_g1k_v37.fasta.gz.

RE: region join, I wouldn't worry about that one too much for now. My thought is that it'd probably be preferable to reimplement vs. the `ReferenceFile` interface instead of the `RegionJoin` interface. cc @laserson for thoughts.

fnothaft commented:
It looks like this works on the WGS data! I did get an exception, but that was unrelated to the MD tag code; the reference file that I provided was missing the decoy contig...

Anywho, this took about 1 hr to run on 16 16 core machines on Spark 1.3.1 with the following Spark options:

```
--master yarn-cluster --num-executors 1024 --executor-memory 11g --executor-cores 1 --driver-memory 14g --conf spark.yarn.executor.memoryOverhead=3072 --conf spark.kryoserializer.buffer.max.mb=1024 --conf spark.driver.maxResultSize=0
```

I'm running again with the decoy contigs from [here](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz) and will report back in the AM.

ryan-williams commented:
Wow, cool, thanks Frank. How big is that BAM, if you know? I'll check tmrw
of not, nbd.

On Fri, Aug 21, 2015, 12:19 AM Frank Austin Nothaft <
notifications@github.com> wrote:

> It looks like this works on the WGS data! I did get an exception, but that
> was unrelated to the MD tag code; the reference file that I provided was
> missing the decoy contig...
>
> Anywho, this took about 1 hr to run on 16 16 core machines on Spark 1.3.1
> with the following Spark options:
>
> --master yarn-cluster --num-executors 1024 --executor-memory 11g --executor-cores 1 --driver-memory 14g --conf spark.yarn.executor.memoryOverhead=3072 --conf spark.kryoserializer.buffer.max.mb=1024 --conf spark.driver.maxResultSize=0
>
> I'm running again with the decoy contigs from here and will report back in
> the AM.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bigdatagenomics/adam/pull/790#issuecomment-133279177>.
>

fnothaft commented:
BAM is 234GB --> approx. 185GB in ADAM.

fnothaft commented:
Completed in approx 2.1 hrs:

<img width="1224" alt="screen shot 2015-08-21 at 6 15 08 am" src="https://cloud.githubusercontent.com/assets/3752466/9409015/07a989b6-47cc-11e5-94a9-55d2041d6078.png">

ryan-williams commented:
Just pushed a commit that should fix this for Spark 1.2.1.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/857/
Test PASSed.

ryan-williams commented:
I'm going to add a couple more things here:
* Option to overwrite existing MDTags (e.g. when you think they may have been incorrectly generated; @arahuja just ran into this).
* Option to use `BroadcastRegionJoin` vs. `ShuffleRegionJoin`, with cmdline flag.
* Option to control the fragment length from the cmdline.

fnothaft commented:
OOC, do you think it might make sense to break this into a separate command?

ryan-williams commented:
Another weird thing I saw when I tested it is that the accumulator values were exactly 3x what they should have been; I had 1000 reads in bqsr.sam, 992 mapped and 8 unmapped, and the accumulators showed up as 2976 "MDTags Added", 24 "Unmapped Reads". Can you spot-check the accumulators' values in your test runs @fnothaft? You should just be able to go to the "saveAsNewAPIHadoopFile at AlignmentRecordRDDFunctions.scala:179" stage page in Spark UI or Spree :)

ryan-williams commented:
re: separate command: maybe!

I was meaning to ask you, in reference to #741, whether you think we are moving toward or away from having commands for `bam2adam`, `adam2bam`, `vcf2adam`, `adam2vcf`, `fasta2adam`, `adam2fasta`. Not necessarily the same thing, but somewhat related to how much we want `transform` to be a kitchen-sink.

fnothaft commented:
Looks like we've got 100% concordance on the big BAM file, with the important aside that it doesn't seem to have MD tags on a large number of the reads.

ryan-williams commented:
How are you checking that OOC? In the description, I showed a pretty hacky thing I did in the shell to just compare the `MD:Z:` lines of two files.

fnothaft commented:
@ryan-williams yeah, I'm not sure. The `transform` kitchen sink is admittedly kind of nice...

fnothaft commented:
```
val rdd1 = sc.loadAlignments("hdfs://amp-bdg-master:8020/user/fnothaft/NA12878_high_coverage.adam")
val rdd2 = sc.loadAlignments("hdfs://amp-bdg-master:8020/user/fnothaft/NA12878_high_coverage.2.adam")
val zRdd = rdd1.zip(rdd2)
zRdd.map(p => { if (p._1.getMismatchingPositions != null && p._1.getMismatchingPositions == p._2.getMismatchingPositions) (1, 1) else (1, 0)}).reduce((v1: (Int, Int), v2: (Int, Int)) => (v1._1 + v2._1, v1._2 + v2._2))
```

ryan-williams commented:
Nice!

Any word on the accumulator values you're seeing?

fnothaft commented:
> Any word on the accumulator values you're seeing?

Ah, I didn't note those when I ran. Do you print those to the log? I'm in a weird network situation right now and can't get to my logs, but I could check those later...

ryan-williams commented:
They end up in the eventlog, if you're writing that; you can fire up a Spark history server that reads it and go to the stage page.

You could also just give me the event log and I'll look.

fnothaft commented:
I believe I should have that. I don't have access to the history server right now (aforementioned weird network connectivity issues...) but will check it when I am able to get access back.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/860/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit a2d9ec92990cc379a7b2be4880ac16b25bf89670.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse a2d9ec92990cc379a7b2be4880ac16b25bf89670^{commit} # timeout=10 > git branch -a --contains a2d9ec92990cc379a7b2be4880ac16b25bf89670 # timeout=10 > git rev-parse remotes/origin/pr/790/head^{commit} # timeout=10Checking out Revision a2d9ec92990cc379a7b2be4880ac16b25bf89670 (origin/pr/790/head) > git config core.sparsecheckout # timeout=10 > git checkout -f a2d9ec92990cc379a7b2be4880ac16b25bf89670First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
Just pushed a big refactor. Still going to work on the `ReferenceFile` stuff, but here's what's new:

* can do a shuffle- or broadcast-based join.
  * haven't tested these head-to-head yet, but will do shortly
* pulled code out to `MDTagging.scala`
* changed the `-bqsr_strict` option to just `-stringency`, so that it can be reused across different `transform` code paths (e.g. for MDTagging here).
* added a `AlignmentRecord => Option[ReferenceRegion]` helper, `ReferenceRegion.opt`; used it in a few places.

ryan-williams commented:
jenkins, retest this please

ryan-williams commented:
I forget if I'm capable of telling Jenkins that, whether it will do it on its own in this case because I've pushed new code, etc.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/861/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 10739bdb5858d2712a65f165326838a57295b500 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 98614dc550c9fd05e606039b1615d1a705b37f31 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 98614dc550c9fd05e606039b1615d1a705b37f31 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 98614dc550c9fd05e606039b1615d1a705b37f31First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/862/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit c398906fb931dd8b392591094014a540d63f2d5a automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 154e8362447882d62ffbac7dc8935a2095d29d48 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 154e8362447882d62ffbac7dc8935a2095d29d48 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 154e8362447882d62ffbac7dc8935a2095d29d48 > git rev-list 98614dc550c9fd05e606039b1615d1a705b37f31 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
jenkins, retest this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/863/<h2>Build result: FAILURE</span></h2>[...truncated 16 lines...] > git rev-list 154e8362447882d62ffbac7dc8935a2095d29d48 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.10,1.2.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.2.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.3.1,centosADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/864/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 37a1f458a89a5c7c92d8e8c373c2f6993b18cfa8 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 5f5688589a785404e08cac9a60d4a5c68eccd7f3 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 5f5688589a785404e08cac9a60d4a5c68eccd7f3 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 5f5688589a785404e08cac9a60d4a5c68eccd7f3 > git rev-list 154e8362447882d62ffbac7dc8935a2095d29d48 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
Jenkins, retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/865/
Test PASSed.

ryan-williams commented:
btw: I figured out why my accumulator values were 3x what they should have been by grepping for them in my event log: I was not caching the RDD where they were computed, and three different stages were computing that RDD. A properly-placed `.cache` fixes it.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/868/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 6d8e218b2bc58652ebec1b2246ff74c0333876b8 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains c8df83b4338eb439314c5968f24eb8604a03c568 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision c8df83b4338eb439314c5968f24eb8604a03c568 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f c8df83b4338eb439314c5968f24eb8604a03c568 > git rev-list 5f5688589a785404e08cac9a60d4a5c68eccd7f3 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/869/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit a53924213652ce6056c4e34ed4ba09250a87a5ca automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 154b540e61733f42231364a7c3d2fc14b18c8a93 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 154b540e61733f42231364a7c3d2fc14b18c8a93 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 154b540e61733f42231364a7c3d2fc14b18c8a93 > git rev-list c8df83b4338eb439314c5968f24eb8604a03c568 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
jenkins, retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/870/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/874/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 35dbfcfe6d51711f92e7c80ee670ecd4f8e8e954 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 2d0e9706456fb9184b3f9455f9423440e936ad75 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 2d0e9706456fb9184b3f9455f9423440e936ad75 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 2d0e9706456fb9184b3f9455f9423440e936ad75 > git rev-list 154b540e61733f42231364a7c3d2fc14b18c8a93 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/875/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit b5ab6ed19873c2ed3d61a1f8199b2f349ec7db52 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 > git rev-list 2d0e9706456fb9184b3f9455f9423440e936ad75 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
jenkins, retest this please

ryan-williams commented:
OK, I took out the shuffle impl and shored up a broadcast impl around a `ReferenceFile` interface. I think it's a pretty good abstraction and we now have N=2 implementations of it so the API is starting to make sense.

I had to add kryo ser/de for `TwoBitFile`.

FWIW, here are Spree screengrabs of the previous broadcast- and shuffle impls on bqsr1.sam:

Broadcast:
![](http://cl.ly/image/0j231O120j2m/Screen%20Shot%202015-08-23%20at%2011.41.39%20AM.png)

Shuffle:
![](http://cl.ly/image/1s0v2n472D2S/Screen%20Shot%202015-08-23%20at%2011.44.49%20AM.png)

The shuffle impl does an extra ~400MB of shuffle read+write, which is interesting. Given that these data sets will basically always be small enough to broadcast, I'm happy with just leaving that impl in.

Here are /jobs pages screenshots of the above, as well:

Broadcast:
![](http://cl.ly/image/2m1Y33453H3q/Screen%20Shot%202015-08-24%20at%202.00.58%20AM.png)

Shuffle:
![](http://cl.ly/image/3I0n190H2d2n/Screen%20Shot%202015-08-24%20at%202.01.19%20AM.png)

Pretty amusing that we shuffle 400MB during the ShuffleRegionJoin; this test was on a 50MB FASTA file and the 1000-read bqsr1.sam.

Anyway, I think this PR is ready to go from my end.

ryan-williams commented:
Ah, one thing I should maybe still do is [let it read `.2bit`s from S3](https://github.com/bigdatagenomics/adam/commit/b5ab6ed19873c2ed3d61a1f8199b2f349ec7db52#diff-d36ea7d0742decd0b040a73a96af06e9R526); I don't know how to do that otomh, pointers welcome!

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/876/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit b5ab6ed19873c2ed3d61a1f8199b2f349ec7db52 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 > git rev-list 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/877/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 6d5b8411ed15181a13d67baf42eaa595b2e5cfd2 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 > git rev-list 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
Retest this please
On Mon, Aug 24, 2015 at 2:19 AM UCB AMPLab <notifications@github.com> wrote:

> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
>
> https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/877/
> Build result: FAILUREGitHub pull request #790 of commit
> 6d5b8411ed15181a13d67baf42eaa595b2e5cfd2 automatically merged.Notifying
> endpoint 'HTTP:
> https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading
> node environment variables.Building remotely on amp-jenkins-worker-05
> (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git
> rev-parse --is-inside-work-tree # timeout=10Fetching changes from the
> remote Git repository > git config remote.origin.url
> https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream
> changes from https://github.com/bigdatagenomics/adam.git > git --version
> # timeout=10 > git fetc h --tags --progress
> https://github.com/bigdatagenomics/adam.git +refs/pull/
> *:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} #
> timeout=10 > git branch -a --contains
> 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 # timeout=10 > git rev-parse
> remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision
> 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 (origin/pr/790/merge) > git config
> core.sparsecheckout # timeout=10 > git checkout -f
> 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 > git rev-list
> 48ba3722426f8aebb4fc9c8c3c0bbd17c5dc2e01 # timeout=10Triggering ADAM-prb ?
> 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ?
> 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so
> aborting...Notifying endpoint 'HTTP:
> https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
> Test FAILed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bigdatagenomics/adam/pull/790#issuecomment-134053086>.
>

ryan-williams commented:
Jenkins, retest this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/878/<h2>Build result: FAILURE</span></h2>[...truncated 16 lines...] > git rev-list 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.10,1.2.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.2.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.3.1,centosADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.11,1.2.1,centos completed with result FAILUREADAM-prb ? 1.0.4,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/879/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 052522bd14d6dbfb59fc411b7cb6c762a6fcc55f automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 647c86007e5198eb965be2215dd47a2130064c15 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 647c86007e5198eb965be2215dd47a2130064c15 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 647c86007e5198eb965be2215dd47a2130064c15 > git rev-list 06fb2cfbb5f1fb44b3cbf86c01b595ba715ce6b4 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
jenkins, retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/880/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/902/<h2>Build result: FAILURE</span></h2>[...truncated 16 lines...] > git rev-list 647c86007e5198eb965be2215dd47a2130064c15 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.3.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.11,1.2.1,centosTriggering ADAM-prb ? 1.0.4,2.10,1.2.1,centosTriggering ADAM-prb ? 2.3.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.3.0,2.10,1.2.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.3.1,centosADAM-prb ? 2.3.0,2.10,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.6.0,2.11,1.3.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.2.1,centos completed with result SUCCESSADAM-prb ? 1.0.4,2.11,1.2.1,centos completed with result SUCCESSADAM-prb ? 1.0.4,2.10,1.2.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.11,1.4.1,centos completed with result SUCCESSADAM-prb ? 2.3.0,2.10,1.2.1,centos completed with result FAILUREADAM-prb ? 2.6.0,2.10,1.3.1,centos completed with result SUCCESSNotifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/903/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit a1e705ef4b10437b3c117db0f7dc984743b02598 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 2e306b3f4cca2d8f26b96aab55a3ce56809227a5 # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 2e306b3f4cca2d8f26b96aab55a3ce56809227a5 (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 2e306b3f4cca2d8f26b96aab55a3ce56809227a5 > git rev-list ff651455f3cfe7603afce2028a5cd0d5689d2078 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/904/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 1f797ee92dada67ca27ee11b7c5936fbec18df04 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains dc629f6e9e95c77859f585b43028cb9369fe592a # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision dc629f6e9e95c77859f585b43028cb9369fe592a (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f dc629f6e9e95c77859f585b43028cb9369fe592a > git rev-list 2e306b3f4cca2d8f26b96aab55a3ce56809227a5 # timeout=10Triggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
jenkins, retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/905/
Test PASSed.

massie commented:
This looks ready to merge to me. What you do you think @fnothaft? I'm happy to do this manually myself, if you think it's ready to go.

fnothaft commented:
This is good to go from my side. @ryan-williams is this ready for merge, or do you have remaining work?

ryan-williams commented:
Yea I just got epically blocked on the probably tiny task of figuring out whether I have to do something special to load files from s3 [here](https://github.com/ryan-williams/adam/commit/1f797ee92dada67ca27ee11b7c5936fbec18df04#diff-d36ea7d0742decd0b040a73a96af06e9R545); I'll defer to y'all on whether that's important for this PR or not!

fnothaft commented:
Oh, hah! I would say let's not worry about that too much for now. We can add it in a follow-on PR.

fnothaft commented:
@ryan-williams Can you rebase this to clean up the merge conflicts? I would like to get this into 0.18.0.

ryan-williams commented:
Working on this now

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/966/<h2>Build result: FAILURE</span></h2>GitHub pull request #790 of commit 72407e56fd91adac83e767c8e0c976eba174d6e2 automatically merged.Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'[EnvInject] - Loading node environment variables.Building remotely on amp-jenkins-worker-05 (centos spark-test) in workspace /home/jenkins/workspace/ADAM-prb > git rev-parse --is-inside-work-tree # timeout=10Fetching changes from the remote Git repository > git config remote.origin.url https://github.com/bigdatagenomics/adam.git # timeout=10Fetching upstream changes from https://github.com/bigdatagenomics/adam.git > git --version # timeout=10 > git fetch --tags --progress https://github.com/bigdatagenomics/adam.git +refs/pull/*:refs/remotes/origin/pr/* > git rev-parse origin/pr/790/merge^{commit} # timeout=10 > git branch -a --contains 53850a892445892400a7616cdee54c4042b8dadc # timeout=10 > git rev-parse remotes/origin/pr/790/merge^{commit} # timeout=10Checking out Revision 53850a892445892400a7616cdee54c4042b8dadc (origin/pr/790/merge) > git config core.sparsecheckout # timeout=10 > git checkout -f 53850a892445892400a7616cdee54c4042b8dadc > git rev-list dc629f6e9e95c77859f585b43028cb9369fe592a # timeout=10First time build. Skipping changelog.Triggering ADAM-prb ? 2.6.0,2.11,1.4.1,centosTriggering ADAM-prb ? 2.6.0,2.10,1.4.1,centosTouchstone configurations resulted in FAILURE, so aborting...Notifying endpoint 'HTTP:https://webhooks.gitter.im/e/ac8bb6e9f53357bc8aa8'
Test FAILed.

ryan-williams commented:
retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/ADAM-prb/967/
Test PASSed.

fnothaft commented:
Thanks @ryan-williams! I've squashed this down and merged as https://github.com/bigdatagenomics/adam/commit/850acdeb5b81cdddd13c1edb06004eb98271823e.
